# Learning Pandas here

In [1]:
import pandas as pd

## Loading CSV data

Pandas gives a great way to consume data from different format like CSV, JSON etc.

In [7]:
df = pd.read_csv('../../data/stackoverflow/survey_results_public.csv')
df.head()

ResponseId     Q120                      MainBranch              Age  \
0           1  I agree                   None of these  18-24 years old   
1           2  I agree  I am a developer by profession  25-34 years old   
2           3  I agree  I am a developer by profession  45-54 years old   
3           4  I agree  I am a developer by profession  25-34 years old   
4           5  I agree  I am a developer by profession  25-34 years old   

                                          Employment  \
0                                                NaN   
1                                Employed, full-time   
2                                Employed, full-time   
3                                Employed, full-time   
4  Employed, full-time;Independent contractor, fr...   

                             RemoteWork  \
0                                   NaN   
1                                Remote   
2  Hybrid (some remote, some in-person)   
3  Hybrid (some remote, some in-person)   
4                                Remote   

                                    CodingActivities  \
0                                                NaN   
1  Hobby;Contribute to open-source projects;Boots...   
2  Hobby;Professional development or self-paced l...   
3                                              Hobby   
4  Hobby;Contribute to open-source projects;Profe...   

                                        EdLevel  \
0                                           NaN   
1  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
2  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
3  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
4  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   

                                           LearnCode  \
0                                                NaN   
1  Books / Physical media;Colleague;Friend or fam...   
2  Books / Physical media;Colleague;On the job tr...   
3  Colleague;Friend or family member;Other online...   
4  Books / Physical media;Online Courses or Certi...   

                                     LearnCodeOnline  ...        Frequency_1  \
0                                                NaN  ...                NaN   
1  Formal documentation provided by the owner of ...  ...   1-2 times a week   
2  Formal documentation provided by the owner of ...  ...  6-10 times a week   
3  Formal documentation provided by the owner of ...  ...   1-2 times a week   
4  Formal documentation provided by the owner of ...  ...   1-2 times a week   

         Frequency_2       Frequency_3         TimeSearching  \
0                NaN               NaN                   NaN   
1   10+ times a week             Never   15-30 minutes a day   
2  6-10 times a week  3-5 times a week   30-60 minutes a day   
3   10+ times a week  1-2 times a week   15-30 minutes a day   
4   1-2 times a week  3-5 times a week  60-120 minutes a day   

         TimeAnswering                                   ProfessionalTech  \
0                  NaN                                                NaN   
1  15-30 minutes a day  DevOps function;Microservices;Automated testin...   
2  30-60 minutes a day  DevOps function;Microservices;Automated testin...   
3  30-60 minutes a day  Automated testing;Continuous integration (CI) ...   
4  30-60 minutes a day  Microservices;Automated testing;Observability ...   

                                            Industry           SurveyLength  \
0                                                NaN                    NaN   
1  Information Services, IT, Software Development...  Appropriate in length   
2  Information Services, IT, Software Development...  Appropriate in length   
3                                                NaN  Appropriate in length   
4                                              Other  Appropriate in length   

                   SurveyEase ConvertedCompYearly  
0                         NaN                 NaN  
1                        Easy            285000.0  
2              

## Getting a sense of the data

There are different way to know about the data.

In [8]:
df.shape

(89184, 84)

So, we can see that this CSV has 

1. Rows: 89,184
2. Columns: 84

When we did a head on the dataframe, we can only see 20 columsn, because by default it gets trimmed.

Even with this, I am not able to view all the columns.

So, I will confifure the head to show all.

In [11]:
pd.set_option('display.max_columns', 84)
df.head()

ResponseId     Q120                      MainBranch              Age  \
0           1  I agree                   None of these  18-24 years old   
1           2  I agree  I am a developer by profession  25-34 years old   
2           3  I agree  I am a developer by profession  45-54 years old   
3           4  I agree  I am a developer by profession  25-34 years old   
4           5  I agree  I am a developer by profession  25-34 years old   

                                          Employment  \
0                                                NaN   
1                                Employed, full-time   
2                                Employed, full-time   
3                                Employed, full-time   
4  Employed, full-time;Independent contractor, fr...   

                             RemoteWork  \
0                                   NaN   
1                                Remote   
2  Hybrid (some remote, some in-person)   
3  Hybrid (some remote, some in-person)   
4                                Remote   

                                    CodingActivities  \
0                                                NaN   
1  Hobby;Contribute to open-source projects;Boots...   
2  Hobby;Professional development or self-paced l...   
3                                              Hobby   
4  Hobby;Contribute to open-source projects;Profe...   

                                        EdLevel  \
0                                           NaN   
1  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
2  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
3  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
4  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   

                                           LearnCode  \
0                                                NaN   
1  Books / Physical media;Colleague;Friend or fam...   
2  Books / Physical media;Colleague;On the job tr...   
3  Colleague;Friend or family member;Other online...   
4  Books / Physical media;Online Courses or Certi...   

                                     LearnCodeOnline  LearnCodeCoursesCert  \
0                                                NaN                   NaN   
1  Formal documentation provided by the owner of ...                 Other   
2  Formal documentation provided by the owner of ...                   NaN   
3  Formal documentation provided by the owner of ...                   NaN   
4  Formal documentation provided by the owner of ...  Other;Codecademy;edX   

  YearsCode YearsCodePro                               DevType  \
0       NaN          NaN                                   NaN   
1        18            9  Senior Executive (C-Suite, VP, etc.)   
2        27           23                   Developer, back-end   
3        12            7                  Developer, front-end   
4         6            4                 Developer, full-stack   

                    OrgSize                 PurchaseInfluence      TechList  \
0                       NaN                               NaN           NaN   
1          2 to 9 employees  I have a great deal of influence   Investigate   
2  5,000 to 9,999 employees             I have some influence  Given a list   
3      100 to 499 employees             I have some influence   Investigate   
4        20 to 99 employees             I have some influence   Investigate   

                                          BuyNewTool  \
0                                                NaN   
1  Start a free trial;Ask developers I know/work ...   
2  Start a free trial;Ask developers I know/work ...   
3  Start a free trial;Ask developers I know/work ...   
4  Start a free trial;Ask developers I know/work ...   

                    Country                   Currency  CompTotal  \
0                       NaN                        NaN        NaN   
1  United States of America  USD\tUnited States dollar   285000.0   
2  United States of America  USD\tUnited States dollar   250000.0   
3  United States of A

Great, now I am able to view the entire data.

It's time to load the schema and see the questions.

So, I will load the other CSV file that I for in the zip file.

In [13]:
schema_df = pd.read_csv('../../data/stackoverflow/survey_results_schema.csv')
schema_df.head()

qid     qname                                           question  \
0   QID16        S0  <div><span style="font-size:19px;"><strong>Hel...   
1   QID12  MetaInfo                                  Browser Meta Info   
2  QID310      Q310  <div><span style="font-size:19px;"><strong>You...   
3  QID312      Q120                                                      
4    QID1        S1  <span style="font-size:22px; font-family: aria...   

  force_resp  type selector  
0      False    DB       TB  
1      False  Meta  Browser  
2      False    DB       TB  
3       True    MC     SAVR  
4      False    DB       TB

I am able to load the data. However, I can see that the question has HTML as well.